documentation/milestones/individual_milestones/developer_tools.md

```diff
@@ -0,0 +1,221 @@
+  - [ ] At least 5 internal developers/maintainers have adopted and run these tools on a cloned version of the repo
+  - [ ] At least 3 open source contributors have adopted and run these tools on a forked version of the repo
+  - [ ] ADRs documenting the selection of these tools and standards have been created and approved
+  - [ ] The status of key checks are reflected as badges in a project README
+- [ ] **Automated testing framework** is live and meets the following conditions:
+  - [ ] At least 1 unit test has been added to the codebase
+  - [ ] At least 1 integration test has been added to the codebase
+  - [ ] Unit tests are run on every push to the remote GitHub repository
+  - [ ] Integration tests are run at least once before merging new code into `main`
+  - [ ] Code which fails any of the unit or integration tests will be blocked from merging into `main`
+  - [ ] A report on the percentage of code covered by tests is available after every test run
+  - [ ] Code whose test coverage falls below a certain threshold will be blocked from merging into `main`
+- [ ] **Code quality checks** are live and meet the following conditions:
+  - [ ] All checks are run on every push to the remote GitHub repository
+  - [ ] The most important checks are run on every local commit
+  - [ ] Code which fails any of these checks will be blocked from merging into `main`
+- [ ] **Security checks** are live and meet the following conditions:
+  - [ ] At least 1 (test) security vulnerability report has been submitted
+  - [ ] Maintainers are notified within 1 hour of a vulnerability being reported within the grants API codebase
+  - [ ] Maintainers are notified within 72 hours of a vulnerability being reported on an upstream dependency
+  - [ ] Security checks are running on every push to the remote GitHub repository
+  - [ ] Code which fails any of these security checks is blocked from merging into `main`
+- [ ] **Dependency management** is live and meets the following conditions:
+  - [ ] At least 2 upstream dependencies have been added to the project
+  - [ ] Compatible versions of upstream dependencies are automatically detected when a new dependency is added
+  - [ ] Maintainers are notified when a new minor or major version of an upstream dependency is available
+- [ ] **Config & secrets management** is live and meets the following conditions:
+  - [ ] At least 1 configuration variable has been added to the project
+  - [ ] This config variable has different values when it is running in different environments (e.g. dev, staging, and prod)
+  - [ ] At least 2 internal developers can access a shared value of this variable locally
+  - [ ] Developers can change the value of this config variable in their local environment without changing it for other developers
+  - [ ] An integration test has been created to validate that this variable is injected at runtime
+
+### Proposed metrics for measuring goals/value/definition of done
+
+- Test framework metrics
```

Comment:
I think developer tools have a couple main goals: reducing bugs, and making the daily experience of software developers more joyful and effective. 

With the former, maybe we can measure and track the number of bug tickets created, perhaps as a percent of all tickets. (I'm guessing we'll probably label bug tickets somehow, e.g., with the label `bug`). Idk, I'm not attached to that metric, I'm just brainstorming.

Maybe we can measure the latter goal somehow too -- like a pulse survey that runs once a month/quarter asking internal and external developers about the enjoyability of working with the codebase. 

I've definitely worked on a few projects where it started out really fun and clean to work in the codebase, and gradually it became annoying as tech debt and weird patterns piled up.

Reply:
I like this! I simplified the metrics to the following:

```
- Total runtime of all CI/CD checks
- Number of bug tickets opened in the last 30 days
- Number of days it takes to close a bug ticket
- Percentage of code covered by unit tests
- Percentage of code covered by integration tests
```

But it could be good to revisit the developer satisfaction idea once we have a better measurement strategy set up. Perhaps we could even do a slack poll or incorporate something like this in our retros?

Reply:
Good ideas!